packages/dds/tree/src/shared-tree/treeCheckout.ts

```diff
@@ -381,20 +403,38 @@ export class TreeCheckout implements ITreeCheckoutFork {
 			idCompressor,
 		),
 	) {
+		// when a transaction is started, take a snapshot of the current state of removed roots
+		transaction.on("transactionStarted", () => {
+			this.removedRootsSnapshots.push(this.removedRoots.clone());
+		});
+		// when a transaction is committed, the latest snapshot of removed roots can be discarded
+		transaction.on("transactionCommitted", () => {
+			this.removedRootsSnapshots.pop();
+		});
+		// when a transaction is aborted, revert removed roots back to the latest snapshot
+		transaction.on("transactionAborted", () => {
+			const snapshot = this.removedRootsSnapshots.pop();
+			assert(snapshot !== undefined, "a snapshot for removed roots does not exist");
+			this.removedRoots.loadIndex(snapshot);
```

Comment:
Instead of doing a deep copy, can you just do this:
```suggestion
			this.removedRoots = snapshot;
```
You'll have to remove the `readonly` modifier, but that seems like a small price to pay here.